### PR TITLE
Fix capping issue

### PR DESCRIPTION
**Pull request by tiffanymatthe**
<s>When tiling, we look at all contours on a specific plane and the contours on the plane below it. If there are no corresponding contours on the plane below it, it will use close_hole_in_floor to close the mesh. However, due to the direction, it will never close_hole_in_roof for a specific plane if it doesn't correspond to the contours on the plane above it.

The fix:
- keep track of contours which do not have associated upper contours -> save in a list of contours (listA) to close hole in roof at the very end.
- also take into account contour amalgamation -- if a contour is saved in listA, but is then erased and replaced by contour amalgamation (from N-N contour correspondence), we need to replace the contour saved in listA by its amalgamated contour.</s>

Update:
- how tiling between planes work: tiles current plane's contours with lower contours. This will take care of lower plane roof holes, current plane's floor holes, and normal meshing.
- underlying issue is: if a contour plane has no upper plane close enough, then it will never be roof capped since roof capping happens when we move to the next plane (but the next plane will never have this plane as its lower plane since it is too far).

Fix:
- if no close enough upper plane, simply cap all the current contours
- can disregard amalgamated contours since those are made for lower to current contour, not current contour to upper contour.

Test cases:
1. Current plane has an upper plane, but no correspondence between them. Expected: two caps.
image
2. Current plane has two contours (which were amalgamated for the single lower contour) and upper plane is too far. Expected: two caps (with no cap for amalgamated).
image
image
4. Current plane has two contours, and next plane has one contour (which only corresponds to one of them). Expected: 1 cap, 1 mesh.
image
image

### CODE REVIEW

File: src/Operations/ConvertContoursToMeshes.cc [hunk 1]
```diff
@@ -272,20 +272,25 @@ bool ConvertContoursToMeshes(Drover &DICOM_data,
...
-            //}
+            if(h_cp_it != std::cend(ucps)){
+               const auto h_cp_dist = std::abs(m_cp_it->Get_Signed_Distance_To_Point(h_cp_it->R_0));
+               if((1.5 * contour_sep) < h_cp_dist) h_cp_it = std::cend(ucps);
+            }
 
             auto l_cops = locate_contours_on_plane(*l_cp_it);
+            auto h_cops = locate_contours_on_plane(*h_cp_it);
+
+            decltype(l_cops) m_cops_to_roof_cap;
```
Comment (tiffanymatthe) [inline src/Operations/ConvertContoursToMeshes.cc]: m_cops_to_roof_cap = list of reference wrappers to all the contours that need a roof cap (close hole in roof).
amalgamated_contours = list of contours (without reference wrappers) that were created in the for loop.
- need this structure to change the scope of the contours created in the for loop, which then can be properly reference wrappered in m_cops_to_roof_cap

Not sure if there is a better way to do this?

**Comment by misprit7** [comment]
Tried reading through the changes but am still a bit confused, lets talk about this tonight

File: src/Operations/ConvertContoursToMeshes.cc [hunk 2]
```diff
@@ -476,11 +501,34 @@ bool ConvertContoursToMeshes(Drover &DICOM_data,
                 return;
             };
 
+            // removes contours in base_cops which are identical to contours in cops_to_remove
+            // returns list of contours remaining in base_cops, but does not modify base_cops
+            // this is used instead of base_cops.erase(...) since erasing an object with a reference wrapper will erase all references to it,
+            // causing seg faults if other list of contours store the same contour
+            // https://stackoverflow.com/a/2874533
```
Comment (misprit7) [inline src/Operations/ConvertContoursToMeshes.cc]: Wait so is this the stl container problem or the set iterator problem described in the stackoverflow post?

File: src/Operations/ConvertContoursToMeshes.cc (hunk 1 above)
Comment (tiffanymatthe) [reply src/Operations/ConvertContoursToMeshes.cc]: ignore, simplified

File: src/Operations/ConvertContoursToMeshes.cc (hunk 2 above)
Comment (tiffanymatthe) [reply src/Operations/ConvertContoursToMeshes.cc]: outdated, was an stl container problem

File: src/Operations/ConvertContoursToMeshes.cc [hunk 3]
```diff
@@ -272,18 +272,23 @@ bool ConvertContoursToMeshes(Drover &DICOM_data,
 
             // Identify whether there are adjacent planes within the contour spacing on either side.
             auto l_cp_it = std::prev(m_cp_it);
-            //auto h_cp_it = std::next(m_cp_it);
+            auto h_cp_it = std::next(m_cp_it);
+
```
Comment (Renu-R) [inline src/Operations/ConvertContoursToMeshes.cc:276]: You've just repeated the commented out line here

**Review by Renu-R — APPROVED** [review]

**Pull request merged**